### PR TITLE
Minor: Simplify conjunction and disjunction, improve docs

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

N/A

## Rationale for this change

I ran into this code while working on #10291 

## What changes are included in this PR?

1. Simplify implementation of `conjunction` and `disjunction` 
2. Add some more docs with examples

## Are these changes tested?
New doc test and existing CI

### CODE REVIEW

File: datafusion/expr/src/utils.rs [hunk 1]
```diff
@@ -1107,20 +1107,49 @@ fn split_binary_impl<'a>(
 /// assert_eq!(conjunction(split), Some(expr));
 /// ```
 pub fn conjunction(filters: impl IntoIterator<Item = Expr>) -> Option<Expr> {
-    filters.into_iter().reduce(|accum, expr| accum.and(expr))
+    filters.into_iter().reduce(Expr::and)
```
Comment (alamb) [inline datafusion/expr/src/utils.rs:1110]: this is the very minor simplification and while I was in here I figured I would improve the docs too

File: datafusion/expr/src/utils.rs [hunk 2]
```diff
@@ -1107,20 +1107,49 @@ fn split_binary_impl<'a>(
...
+///   col("a").eq(lit(1)),
+///   col("b").eq(lit(2)),
+/// ];
+///
+/// // use disjuncton to join them together with `OR`
+/// assert_eq!(disjunction(split), Some(expr));
+/// ```
 pub fn disjunction(filters: impl IntoIterator<Item = Expr>) -> Option<Expr> {
-    filters.into_iter().reduce(|accum, expr| accum.or(expr))
+    filters.into_iter().reduce(Expr::or)
```
Comment (jayzhan211) [inline datafusion/expr/src/utils.rs:1136]: Wow, surprise that this works

**Review by jayzhan211 — APPROVED** [review]
👍

File: datafusion/expr/src/utils.rs (hunk 2 above)
Comment (alamb) [reply datafusion/expr/src/utils.rs:1136]: Yeah, that was my first reaction when I saw this in Filter pushdown and it took me a while to figure out what was actually doing 🤯

**Pull request merged**